File: grokking_algorithms/go/breadth_first_search.go
package main

import "fmt"

func breadthFirstSearch(graph map[string][]string, name string) bool {
	// var queue []string
	// queue = append(queue, graph[name]...)

	queue := graph[name]

	var visited []string

	for len(queue) != 0 {
		person := queue[0]
		queue = queue[1:]

		if !contains(person, visited) {
			if isSeller(person) {
				fmt.Println("You fund mango seller:", person)
				return true
			}

			queue = append(queue, graph[person]...)
			visited = append(visited, person)
		}
	}

	return false
}

func breadthFirstSearchV2[K comparable](graph map[K][]K, v, t K) bool {
	queue := []K{v}
	visited := []K{}

	for len(queue) != 0 {
		neighbor := queue[0]
		queue = queue[1:]

		if !contains(neighbor, visited) {
			if neighbor == t {
				return true
			}

			queue = append(queue, graph[neighbor]...)
			visited = append(visited, neighbor)
		}
	}

	return false
}

File: grokking_algorithms/rust/src/breadth_first_search.rs
use std::{collections::HashMap, hash::Hash};

#[allow(dead_code)]
fn breadth_first_search<T>(graph: HashMap<T, Vec<T>>, v: T, t: T) -> bool
where
    T: PartialOrd + Eq + Hash + Clone,
{
    let mut queue: Vec<T> = graph.get(&v).unwrap().to_vec();
    let mut visited: Vec<T> = Vec::new();

    while queue.len() != 0 {
        let neighbor = queue.pop().unwrap();

        if !visited.contains(&neighbor) {
            if neighbor == t {
                return true;
            }

            queue.extend(graph.get(&neighbor).unwrap().to_vec());
            visited.push(neighbor);
        }
    }

    false
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_breadth_first_search_success() {
        let graph: HashMap<String, Vec<String>> = [
            ("A", vec!["E", "B", "C"]),
            ("E", vec!["I", "W"]),
            ("B", vec!["W", "D"]),
            ("C", vec!["U"]),
            ("I", vec!["Y", "U"]),
            ("W", vec![]),
            ("D", vec![]),
            ("Y", vec!["O"]),
            ("U", vec!["F"]),
            ("O", vec![]),
            ("F", vec![]),
        ]
        .iter()
        .map(|(k, v)| {
            (
                String::from(*k),
                v.iter().map(|s| String::from(*s)).collect(),
            )
        })
        .collect();

        let result = breadth_first_search(graph, "A".to_string(), "F".to_string());

        assert_eq!(result, true)
    }

    #[test]
    fn test_breadth_first_search_fail() {
        let graph: HashMap<String, Vec<String>> = [
            ("A", vec!["E", "B", "C"]),
            ("E", vec!["I", "W"]),
            ("B", vec!["W", "D"]),
            ("C", vec!["U"]),
            ("I", vec!["Y", "U"]),
            ("W", vec![]),
            ("D", vec![]),
            ("Y", vec!["O"]),
            ("U", vec!["F"]),
            ("O", vec![]),
            ("F", vec![]),
        ]
        .iter()
        .map(|(k, v)| {
            (
                String::from(*k),
                v.iter().map(|s| String::from(*s)).collect(),
            )
        })
        .collect();

        let result = breadth_first_search(graph, "B".to_string(), "F".to_string());

        assert_eq!(result, false)
    }
}
